File: lib/presentation/views/week/week_view.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import '../../models/habit.dart';
import '../../providers/habit_provider.dart';
import '../../services/daily_score_service.dart';
import '../../services/week_cycle_manager.dart';
import '../../theme/app_theme.dart';

class WeekView extends StatelessWidget {
  final WeekCycleManager weekCycleManager;

  const WeekView({super.key, required this.weekCycleManager});

  static const _scoreService = DailyScoreService.instance;
  static const _dayLabels = ['S', 'M', 'T', 'W', 'T', 'F', 'S'];

  @override
  Widget build(BuildContext context) {
    final habits = context.watch<HabitProvider>().sortedHabits;
    final today = DateTime.now();
    final todayStart = DateTime(today.year, today.month, today.day);
    final weekDates = weekCycleManager.currentWeekDates;
    final daysElapsed = weekDates.where((d) => !d.isAfter(todayStart)).toList();

    final overallScore = daysElapsed.isEmpty
        ? 0.0
        : daysElapsed.map((d) => _scoreService.dailyScore(habits.toList(), d)).reduce((a, b) => a + b) /
            daysElapsed.length;
    final tier = _scoreService.tierForScore(overallScore);

    final totalCompleted = habits.fold<int>(0, (sum, h) => sum + weekCycleManager.completedDaysThisWeek(h));
    final totalPossible = daysElapsed.fold<int>(
        0, (sum, d) => sum + habits.where((h) => h.isActive(d)).length);

    final previews = habits
        .map((h) {
          final preview = weekCycleManager.microMilestonePreview(h);
          return preview != null ? (h, preview) : null;
        })
        .whereType<(Habit, String)>()
        .toList();

    return Scaffold(
      backgroundColor: TributeColor.charcoal,
      body: SafeArea(
        child: CustomScrollView(
          slivers: [
            SliverAppBar(
              backgroundColor: TributeColor.charcoal,
              title: const Text('This Week',
                  style: TextStyle(
                      color: TributeColor.warmWhite, fontSize: 22, fontWeight: FontWeight.w700)),
              floating: true,
              snap: true,
              pinned: false,
            ),
            SliverPadding(
              padding: const EdgeInsets.fromLTRB(16, 8, 16, 32),
              sliver: SliverList(
                delegate: SliverChildListDelegate([
                  _weekSummaryHeader(tier, totalCompleted, totalPossible),
                  const SizedBox(height: 20),
                  if (previews.isNotEmpty) ...[
                    _milestoneCallouts(previews),
                    const SizedBox(height: 20),
                  ],
                  ...habits.map((h) => Padding(
                        padding: const EdgeInsets.only(bottom: 12),
                        child: _habitWeekCard(h, weekDates, todayStart),
                      )),
                ]),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _weekSummaryHeader(DayTier tier, int completed, int possible) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const SizedBox(height: 8),
        Text('Your Week So Far',
            style: TextStyle(
                fontSize: 15, fontWeight: FontWeight.w600, color: TributeColor.softGold)),
        const SizedBox(height: 14),
        Row(
          children: [
            SizedBox(width: 52, height: 52, child: _tierIndicator(tier)),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    _tierLabel(tier),
                    style: const TextStyle(
                        fontSize: 18, fontWeight: FontWeight.w700, color: TributeColor.warmWhite),
                  ),
                  Text(
                    weekCycleManager.graceMessage(completed, possible),
                    style: const TextStyle(fontSize: 12, color: TributeColor.sage),
                  ),
                ],
              ),
            ),
          ],
        ),
      ],
    );
  }

  Widget _tierIndicator(DayTier tier) {
    switch (tier) {
      case DayTier.nothing:
        return Container(
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            border: Border.all(color: Colors.white.withValues(alpha: 0.08), width: 1.5),
          ),
        );
      case DayTier.partial:
        return Container(
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            border: Border.all(color: TributeColor.golden.withValues(alpha: 0.6), width: 2),
          ),
          child: Center(
            child: Icon(Icons.check, size: 16, color: TributeColor.golden.withValues(alpha: 0.7)),
          ),
        );
      case DayTier.substantial:
        return Container(
          decoration: const BoxDecoration(shape: BoxShape.circle, color: TributeColor.golden),
          child: const Center(child: Icon(Icons.check, size: 18, color: TributeColor.charcoal)),
        );
      case DayTier.full:
        return Stack(
          alignment: Alignment.center,
          children: [
            Container(
              width: 64,
              height: 64,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                border: Border.all(color: TributeColor.golden.withValues(alpha: 0.45), width: 1.5),
              ),
            ),
            Container(
              width: 52,
              height: 52,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: TributeColor.golden,
                boxShadow: [
                  BoxShadow(color: TributeColor.golden.withValues(alpha: 0.7), blurRadius: 14),
                  BoxShadow(color: TributeColor.golden.withValues(alpha: 0.35), blurRadius: 5),
                ],
              ),
              child: const Center(child: Icon(Icons.check, size: 18, color: TributeColor.charcoal)),
            ),
          ],
        );
    }
  }

  String _tierLabel(DayTier tier) {
    switch (tier) {
      case DayTier.nothing: return 'Just getting started';
      case DayTier.partial: return 'Something given';
      case DayTier.substantial: return 'Strong week';
      case DayTier.full: return 'Beautiful week';
    }
  }

  Widget _milestoneCallouts(List<(Habit, String)> previews) {
    return Container(
      padding: const EdgeInsets.all(14),
      decoration: BoxDecoration(
        color: TributeColor.golden.withValues(alpha: 0.04),
        borderRadius: BorderRadius.circular(14),
        border: Border.all(color: TributeColor.golden.withValues(alpha: 0.12), width: 0.5),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: previews.map((item) {
          final (habit, preview) = item;
          return Padding(
            padding: const EdgeInsets.only(bottom: 8),
            child: Row(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                const Icon(Icons.auto_awesome, size: 11, color: TributeColor.golden),
                const SizedBox(width: 8),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(habit.name,
                          style: const TextStyle(
                              fontSize: 12,
                              fontWeight: FontWeight.w600,
                              color: TributeColor.warmWhite)),
                      Text(preview,
                          style: TextStyle(
                              fontSize: 11,
                              color: TributeColor.softGold.withValues(alpha: 0.7))),
                    ],
                  ),
                ),
              ],
            ),
          );
        }).toList(),
      ),
    );
  }

  Widget _habitWeekCard(Habit habit, List<DateTime> weekDates, DateTime todayStart) {
    final isAbstain = habit.habitTrackingType == HabitTrackingType.abstain;
    final accent = isAbstain ? TributeColor.sage : TributeColor.golden;
    final daysElapsed = weekDates.where((d) => !d.isAfter(todayStart)).toList();
    final milestone = weekCycleManager.microMilestonePreview(habit);

    return Container(
      padding: const EdgeInsets.all(16),
      decoration: TributeDecorations.card,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Icon(_habitIcon(habit), size: 18, color: accent),
              const SizedBox(width: 10),
              Expanded(
                child: Text(habit.name,
                    style: const TextStyle(
                        fontSize: 15,
                        fontWeight: FontWeight.w600,
                        color: TributeColor.warmWhite)),
              ),
              Text(_weekSummaryText(habit, daysElapsed),
                  style: TextStyle(
                      fontSize: 12,
                      fontWeight: FontWeight.w500,
                      color: TributeColor.softGold.withValues(alpha: 0.7))),
            ],
          ),
          const SizedBox(height: 14),
          Row(
            children: weekDates.asMap().entries.map((e) {
              final i = e.key;
              final date = e.value;
              final isFuture = date.isAfter(todayStart);
              final isToday = date.year == todayStart.year &&
                  date.month == todayStart.month &&
                  date.day == todayStart.day;
              final isActive = habit.isActive(date);
              final score = isActive ? _scoreService.habitScore(habit, date) : -1.0;
              final tileTier = _scoreService.tierForScore(score.clamp(0.0, 1.0));

              return Expanded(
                child: Column(
                  children: [
                    Text(_dayLabels[i],
                        style: TextStyle(
                            fontSize: 10,
                            fontWeight: FontWeight.w500,
                            color: isToday
                                ? TributeColor.softGold
                                : Colors.white.withValues(alpha: 0.4))),
                    const SizedBox(height: 6),
                    _habitDayTile(habit, isActive, isFuture, isToday, tileTier, accent),
                  ],
                ),
              );
            }).toList(),
          ),
          if (milestone != null) ...[
            const SizedBox(height: 10),
            Row(
              children: [
                const Icon(Icons.auto_awesome, size: 11, color: TributeColor.golden),
                const SizedBox(width: 6),
                Expanded(
                  child: Text(milestone,
                      style: TextStyle(
                          fontSize: 11,
                          color: TributeColor.softGold.withValues(alpha: 0.6))),
                ),
              ],
            ),
          ],
        ],
      ),
    );
  }

  Widget _habitDayTile(
      Habit habit, bool isActive, bool isFuture, bool isToday, DayTier tier, Color accent) {
    if (!isActive) {
      return SizedBox(
        width: 36,
        height: 36,
        child: Center(
          child: Text('\u2013',
              style: TextStyle(fontSize: 10, color: Colors.white.withValues(alpha: 0.15))),
        ),
      );
    }
    if (isFuture) {
      return Container(
        width: 36,
        height: 36,
        decoration: BoxDecoration(
            shape: BoxShape.circle, color: Colors.white.withValues(alpha: 0.03)),
      );
    }
    final isAbstain = habit.habitTrackingType == HabitTrackingType.abstain;
    final icon = isAbstain ? Icons.shield_rounded : Icons.check;

    switch (tier) {
      case DayTier.nothing:
        return Container(
          width: 36,
          height: 36,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            border: Border.all(
              color: isToday
                  ? TributeColor.golden.withValues(alpha: 0.4)
                  : Colors.white.withValues(alpha: 0.08),
              width: isToday ? 1.5 : 1,
            ),
          ),
        );
      case DayTier.partial:
        return Container(
          width: 36,
          height: 36,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            border: Border.all(color: accent.withValues(alpha: 0.6), width: 1.5),
          ),
          child: Center(child: Icon(icon, size: 13, color: accent.withValues(alpha: 0.7))),
        );
      case DayTier.substantial:
        return Container(
          width: 36,
          height: 36,
          decoration: BoxDecoration(shape: BoxShape.circle, color: accent),
          child: Center(child: Icon(icon, size: 14, color: TributeColor.charcoal)),
        );
      case DayTier.full:
        return Stack(
          alignment: Alignment.center,
          children: [
            Container(
              width: 44,
              height: 44,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                border: Border.all(color: accent.withValues(alpha: 0.5), width: 1.5),
              ),
            ),
            Container(
              width: 36,
              height: 36,
              decoration: BoxDecoration(
                shape: BoxShape.circle,
                color: accent,
                boxShadow: [BoxShadow(color: accent.withValues(alpha: 0.7), blurRadius: 10)],
              ),
              child: Center(child: Icon(icon, size: 14, color: TributeColor.charcoal)),
            ),
          ],
        );
    }
  }

  String _weekSummaryText(Habit habit, List<DateTime> daysElapsed) {
    final activeDays = daysElapsed.where((d) => habit.isActive(d)).toList();
    final count = activeDays.length;

    switch (habit.habitTrackingType) {
      case HabitTrackingType.timed:
        final total = activeDays.fold<double>(0, (s, d) => s + (habit.entryFor(d)?.value ?? 0));
        final target = habit.dailyTarget * count;
        return '${total.toInt()} / ${target.toInt()} min';
      case HabitTrackingType.count:
        final total = activeDays.fold<double>(0, (s, d) => s + (habit.entryFor(d)?.value ?? 0));
        final target = habit.dailyTarget * count;
        final unit = habit.targetUnit.isEmpty ? '' : ' ${habit.targetUnit}';
        return '${total.toInt()} / ${target.toInt()}$unit';
      case HabitTrackingType.checkIn:
        final done = activeDays.where((d) => habit.isCompleted(d)).length;
        return '$done / $count days';
      case HabitTrackingType.abstain:
        final done = activeDays.where((d) => habit.isCompleted(d)).length;
        return '$done / $count days clean';
    }
  }

  IconData _habitIcon(Habit habit) {
    if (habit.habitTrackingType == HabitTrackingType.abstain) return Icons.shield_rounded;
    switch (habit.habitCategory) {
      case HabitCategory.gratitude: return Icons.auto_awesome;
      case HabitCategory.scripture: return Icons.menu_book;
      case HabitCategory.exercise: return Icons.fitness_center;
      case HabitCategory.rest: return Icons.bedtime;
      case HabitCategory.fasting: return Icons.no_food;
      case HabitCategory.study: return Icons.school;
      case HabitCategory.service: return Icons.volunteer_activism;
      case HabitCategory.connection: return Icons.people;
      case HabitCategory.health: return Icons.favorite;
      case HabitCategory.abstain: return Icons.shield_rounded;
      case HabitCategory.custom: return Icons.star;
    }
  }
}

File: swift_reference/Views/WeekView.swift
import SwiftUI
import SwiftData

struct WeekView: View {
    let weekCycleManager: WeekCycleManager

    @Query(sort: \Habit.sortOrder) private var habits: [Habit]
    @State private var scoreService = DailyScoreService()

    private let dayLabels = ["S", "M", "T", "W", "T", "F", "S"]
    private let calendar = Calendar.current

    private var weekDates: [Date] {
        weekCycleManager.currentWeekDates
    }

    private var todayStart: Date {
        calendar.startOfDay(for: Date())
    }

    private var daysElapsed: [Date] {
        weekDates.filter { calendar.startOfDay(for: $0) <= todayStart }
    }

    private var overallWeekScore: Double {
        guard !daysElapsed.isEmpty else { return 0 }
        let dayScores = daysElapsed.map { scoreService.dailyScore(for: habits, on: $0) }
        return dayScores.reduce(0, +) / Double(dayScores.count)
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: 24) {
                    weekSummaryHeader
                    milestoneCallouts

                    ForEach(habits) { habit in
                        habitWeekCard(habit)
                    }
                }
                .padding(.horizontal, 16)
                .padding(.bottom, 32)
            }
            .background {
                ZStack {
                    TributeColor.charcoal.ignoresSafeArea()
                    TributeColor.warmGlow.ignoresSafeArea()
                }
            }
            .navigationTitle("This Week")
            .navigationBarTitleDisplayMode(.large)
            .toolbarColorScheme(.dark, for: .navigationBar)
        }
    }

    private var weekSummaryHeader: some View {
        VStack(alignment: .leading, spacing: 14) {
            HStack {
                Text("Your Week So Far")
                    .font(.system(.headline, design: .serif))
                    .foregroundStyle(TributeColor.softGold)

                Spacer()

                if weekCycleManager.isCurrentWeekDedicated {
                    HStack(spacing: 4) {
                        Image(systemName: "checkmark.seal.fill")
                            .font(.caption2)
                        Text("Dedicated")
                            .font(.caption2.weight(.medium))
                    }
                    .foregroundStyle(TributeColor.golden)
                    .padding(.horizontal, 8)
                    .padding(.vertical, 4)
                    .background(TributeColor.golden.opacity(0.1))
                    .clipShape(Capsule())
                }
            }

            HStack(spacing: 12) {
                weekTierIndicator
                    .frame(width: 52, height: 52)

                VStack(alignment: .leading, spacing: 4) {
                    Text(weekTierLabel)
                        .font(.system(.title3, design: .serif, weight: .bold))
                        .foregroundStyle(.primary)

                    Text(weekCycleManager.graceMessage(
                        completed: totalCompletedCheckIns,
                        total: totalPossibleCheckIns
                    ))
                    .font(.system(.caption, design: .serif))
                    .foregroundStyle(TributeColor.sage)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.top, 8)
    }

    @ViewBuilder
    private var weekTierIndicator: some View {
        let tier = scoreService.tier(for: overallWeekScore)
        switch tier {
        case .nothing:
            Circle()
                .strokeBorder(Color.white.opacity(0.08), lineWidth: 1.5)
        case .partial:
            Circle()
                .strokeBorder(TributeColor.golden.opacity(0.6), lineWidth: 2)
                .overlay {
                    Image(systemName: "checkmark")
                        .font(.system(size: 14, weight: .bold))
                        .foregroundStyle(TributeColor.golden.opacity(0.7))
                }
        case .substantial:
            Circle()
                .fill(TributeColor.golden)
                .overlay {
                    Image(systemName: "checkmark")
                        .font(.system(size: 16, weight: .bold))
                        .foregroundStyle(TributeColor.charcoal)
                }
        case .full:
            Circle()
                .fill(TributeColor.golden)
                .overlay {
                    Circle()
                        .strokeBorder(TributeColor.golden.opacity(0.45), lineWidth: 2)
                        .scaleEffect(1.3)
                }
                .shadow(color: TributeColor.golden.opacity(0.7), radius: 14, y: 0)
                .shadow(color: TributeColor.golden.opacity(0.35), radius: 5, y: 0)
                .overlay {
                    Image(systemName: "checkmark")
                        .font(.system(size: 16, weight: .bold))
                        .foregroundStyle(TributeColor.charcoal)
                }
        }
    }

    private var weekTierLabel: String {
        let tier = scoreService.tier(for: overallWeekScore)
        switch tier {
        case .nothing: return "Just getting started"
        case .partial: return "Something given"
        case .substantial: return "Strong week"
        case .full: return "Beautiful week"
        }
    }

    private var totalCompletedCheckIns: Int {
        habits.reduce(0) { $0 + weekCycleManager.completedDaysThisWeek(for: $1) }
    }

    private var totalPossibleCheckIns: Int {
        daysElapsed.reduce(0) { total, date in
            total + habits.filter { $0.isActive(on: date) }.count
        }
    }

    @ViewBuilder
    private var milestoneCallouts: some View {
        let previews = habits.compactMap { habit -> (Habit, String)? in
            guard let preview = weekCycleManager.microMilestonePreview(for: habit) else { return nil }
            return (habit, preview)
        }

        if !previews.isEmpty {
            VStack(alignment: .leading, spacing: 10) {
                ForEach(previews, id: \.0.id) { habit, preview in
                    HStack(alignment: .top, spacing: 8) {
                        Image(systemName: "sparkle")
                            .font(.caption2)
                            .foregroundStyle(TributeColor.golden)
                            .padding(.top, 2)

                        VStack(alignment: .leading, spacing: 2) {
                            Text(habit.name)
                                .font(.caption.weight(.semibold))
                                .foregroundStyle(.primary)
                            Text(preview)
                                .font(.caption)
                                .foregroundStyle(TributeColor.softGold.opacity(0.7))
                        }
                    }
                }
            }
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(TributeColor.golden.opacity(0.04))
            .clipShape(.rect(cornerRadius: 14))
            .overlay(
                RoundedRectangle(cornerRadius: 14)
                    .strokeBorder(TributeColor.golden.opacity(0.12), lineWidth: 0.5)
            )
        }
    }

    private func habitWeekCard(_ habit: Habit) -> some View {
        VStack(alignment: .leading, spacing: 14) {
            HStack(spacing: 10) {
                Image(systemName: habit.habitCategory.iconName)
                    .font(.system(size: 16))
                    .foregroundStyle(habit.habitTrackingType == .abstain ? TributeColor.sage : TributeColor.golden)

                Text(habit.name)
                    .font(.system(.headline, design: .serif))

                Spacer()

                Text(weekSummaryText(for: habit))
                    .font(.system(.caption, design: .rounded, weight: .medium))
                    .foregroundStyle(TributeColor.softGold.opacity(0.7))
            }

            HStack(spacing: 6) {
                ForEach(Array(weekDates.enumerated()), id: \.offset) { index, date in
                    let dateStart = calendar.startOfDay(for: date)
                    let isFuture = dateStart > todayStart
                    let isToday = calendar.isDateInToday(date)
                    let isActive = habit.isActive(on: date)
                    let entry = habit.entry(for: date)
                    let isCompleted = entry?.isCompleted ?? false
                    let score = isActive ? scoreService.habitScore(for: habit, on: date) : -1

                    VStack(spacing: 6) {
                        Text(dayLabels[index])
                            .font(.system(.caption2, design: .rounded, weight: .medium))
                            .foregroundStyle(isToday ? TributeColor.softGold : .secondary)

                        ZStack {
                            if !isActive {
                                Circle()
                                    .fill(Color.white.opacity(0.02))
                                    .frame(width: 36, height: 36)
                                Text("\u{2013}")
                                    .font(.caption2)
                                    .foregroundStyle(.secondary.opacity(0.3))
                            } else if isFuture {
                                Circle()
                                    .fill(Color.white.opacity(0.03))
                                    .frame(width: 36, height: 36)
                            } else {
                                habitDayTile(
                                    habit: habit,
                                    isCompleted: isCompleted,
                                    score: max(score, 0),
                                    isToday: isToday,
                                    entry: entry
                                )
                            }
                        }
                    }
                    .frame(maxWidth: .infinity)
                }
            }

            if let milestone = weekCycleManager.microMilestonePreview(for: habit) {
                HStack(spacing: 6) {
                    Image(systemName: "sparkle")
                        .font(.caption2)
                        .foregroundStyle(TributeColor.golden)
                    Text(milestone)
                        .font(.caption)
                        .foregroundStyle(TributeColor.softGold.opacity(0.6))
                }
                .padding(.top, 2)
            }
        }
        .tributeCard()
    }

    @ViewBuilder
    private func habitDayTile(habit: Habit, isCompleted: Bool, score: Double, isToday: Bool, entry: HabitEntry?) -> some View {
        let tier = scoreService.tier(for: score)

        switch tier {
        case .nothing:
            Circle()
                .strokeBorder(Color.white.opacity(0.08), lineWidth: 1)
                .frame(width: 36, height: 36)
                .overlay {
                    if isToday {
                        Circle()
                            .strokeBorder(TributeColor.golden.opacity(0.4), lineWidth: 1.5)
                            .frame(width: 36, height: 36)
                    }
                }

        case .partial:
            Circle()
                .strokeBorder(TributeColor.golden.opacity(0.6), lineWidth: 1.5)
                .frame(width: 36, height: 36)
                .overlay {
                    habitDayIcon(habit: habit, tier: .partial)
                }

        case .substantial:
            Circle()
                .fill(habit.habitTrackingType == .abstain ? TributeColor.sage : TributeColor.golden)
                .frame(width: 36, height: 36)
                .overlay {
                    habitDayIcon(habit: habit, tier: .substantial)
                }

        case .full:
            let accentColor = habit.habitTrackingType == .abstain ? TributeColor.sage : TributeColor.golden
            Circle()
                .fill(accentColor)
                .frame(width: 36, height: 36)
                .overlay {
                    Circle()
                        .strokeBorder(accentColor.opacity(0.5), lineWidth: 1.5)
                        .frame(width: 44, height: 44)
                }
                .shadow(color: accentColor.opacity(0.7), radius: 10, y: 0)
                .shadow(color: accentColor.opacity(0.3), radius: 4, y: 0)
                .overlay {
                    habitDayIcon(habit: habit, tier: .full)
                }
        }
    }

    @ViewBuilder
    private func habitDayIcon(habit: Habit, tier: DayTier) -> some View {
        if habit.habitTrackingType == .abstain {
            Image(systemName: "shield.fill")
                .font(.system(size: 12))
                .foregroundStyle(tier == .partial ? TributeColor.sage.opacity(0.7) : TributeColor.charcoal)
        } else {
            Image(systemName: "checkmark")
                .font(.system(size: 10, weight: .bold))
                .foregroundStyle(tier == .partial ? TributeColor.golden.opacity(0.7) : TributeColor.charcoal)
        }
    }

    private func weekSummaryText(for habit: Habit) -> String {
        let activeDays = daysElapsed.filter { habit.isActive(on: $0) }
        let activeDayCount = activeDays.count

        switch habit.habitTrackingType {
        case .timed:
            let totalMinutes = activeDays.reduce(0.0) { $0 + (habit.entry(for: $1)?.value ?? 0) }
            let targetMinutes = habit.dailyTarget * Double(activeDayCount)
            if targetMinutes >= 60 {
                return String(format: "%.0f / %.0f min", totalMinutes, targetMinutes)
            }
            return "\(Int(totalMinutes)) / \(Int(targetMinutes)) min"

        case .count:
            let totalCount = activeDays.reduce(0.0) { $0 + (habit.entry(for: $1)?.value ?? 0) }
            let targetCount = habit.dailyTarget * Double(activeDayCount)
            let unit = habit.targetUnit.isEmpty ? "" : " \(habit.targetUnit)"
            return "\(Int(totalCount)) / \(Int(targetCount))\(unit)"

        case .checkIn:
            let completed = activeDays.filter { habit.isCompleted(on: $0) }.count
            return "\(completed) / \(activeDayCount) days"

        case .abstain:
            let confirmed = activeDays.filter { habit.isCompleted(on: $0) }.count
            return "\(confirmed) / \(activeDayCount) days clean"
        }
    }
}
